### ISSUE TITLE
Virtual keyboard selects special characters during fast input

### ISSUE THREAD

**Issue by tarkeez** [koreader/koreader#11668]
* KOReader version: 2024.03.1
* Device: Onyx Boox Page (Android 11)

#### Issue
When typing (e.g. adding a note to a highlight), a sequence of two characters (e.g. n + g) will be interpreted a single special character (e.g. ñ) if they are typed in quick succession.

#### Steps to reproduce
Open input field
Have latin based keyboard (e.g. English)
Type two letters in quick succession
Occasionally get a single special character instead of the two characters you typed

#### Possible cause
What sems to happen is that the special character chosen corresponds to the area or relative direction where that special character would be in the pop-up if you long press the base character in question. For n:

![image](https://github.com/koreader/koreader/assets/156511381/ff8cb1c1-c785-4b3e-b077-3a0604b96746)

So n + character to the straight-ish left of n gives ñ, n + character to the upper right of n gives ń, n + straight up gives N etc, you get the point.

**So when typing fast, sometimes it seems to trigger the behavior of long pressing + choosing a special character around the base letter.**

**Comment by hius07** [comment]
Upper menu Gear - Device - Keyboard - Swipe to input additional characters

**Comment by tarkeez** [comment]
Nice -- yes that was it.

Now the issue becomes that when quick successive taps are interpreted as swipes (and thus ignored when the settings you mentioned is off), you get nothing. 

Can anything be done to accommodate us super speed ultra turbo typers? Like skip the check for swipe completely when that setting is off maybe?

**Comment by hius07** [comment]
> quick successive taps are interpreted as swipes

Actually only the first tap is interpreted as swipe and is ignored.
So the idea is, if the setting is off, to handle swipes as taps. Makes sense.

**Comment by tarkeez** [comment]
Fair enough, I guess the second tap becomes the swipe direction. 

But is the only solution here to type slower?

Or can the swipe be disabled entirely (instead of registered but ignored) to allow for very fast typing?

**Comment by hius07** [comment]
Can you try if it is the desired behaviour?

```Lua
local VirtualKeyboard = require("ui/widget/virtualkeyboard")
local VirtualKey
local n = 1
while true do
    local name, value = debug.getupvalue(VirtualKeyboard.addKeys, n)
    if name == "VirtualKey" then
        VirtualKey = value
        break
    end
    n = n + 1
end
VirtualKey.onSwipeKey = VirtualKey.onTapSelect

```
Create a file named `2-key-swipe-to-tap.lua` containing the above lines, put the file into koreader/patches folder, restart KOReader.
(It doesn't depend on the `Swipe to input additional characters` setting)

**Comment by hius07** [comment]
Is it actually a combination "swipe+tap"?
And not just a swipe from one key to another?
I tried to reproduce on my Kindle, with as small interval as I can do, and the second tap was detected. (Double taps are disabled in the settings)

**Comment by tarkeez** [comment]
I am definitely just tapping the keys. I tried disabling double taps, but I'm getting the same results.

If tap three keys in quick succession I get either 1, 12, or 13, never 123. So maybe my device (boox page) has problems with quick successive taps?

I'll try on another Android device when I get the change and post the results.

Thanks for taking the time.

**Comment by offset-torque** [comment]
@hius07

On Libra I tried many times. KOReader caught all of them correctly.
Fastest trial was 4 keys in ~500 ms and still they were all correctly entered.
Looks like some Android weirdness (buffering and merging taps?).

**Comment by pazos** [comment]
> Looks like some Android weirdness (buffering and merging taps?).

Hopefully it isn't and the issue at hand is relegated to a few broken eink/android devices.

On my phone, with haptic feedback enabled, I'm able to type almost as fast as with the regular android keyboard. Okay, not **almost** but it seems related to the lack of training and mismatches with the keyboard layout I'm used to in a phone.

**Comment by hugleo** [comment]
> > Looks like some Android weirdness (buffering and merging taps?).
> 
> Hopefully it isn't and the issue at hand is relegated to a few broken eink/android devices.
> 
> On my phone, with haptic feedback enabled, I'm able to type almost as fast as with the regular android keyboard. Okay, not **almost** but it seems related to the lack of training and mismatches with the keyboard layout I'm used to in a phone.

It isn't, looks normal in my Onyx Ultra C, also Android 11. Unless of course, you got supernatural fingers.

**Comment by ws2bhhh** [comment]
@tarkeez  gboard, zh_cn keyboard has the same problem on eink android phones. I manually setting the Swipe to input sensitivity to low to fix this. Maybe you have a similar setting on your device.

**Comment by tarkeez** [comment]
I did a quick test now on a regular Android device.

Without patch:
The behavior (either ignoring quick successive taps or giving a special character, depending on whether swipe for special character is on or not) is more difficult to trigger (so in practice would not get in the way as much, maybe at all, depending on your typing speed) -- seems I have to input the two letters quicker than on the e-ink boox page to trigger it. 

With patch:
Same behavior as expected. Have to input faster than on the eink device, but the same behavior happens: if the taps are close enough together, the second tap is ignored.

So the e-ink device (maybe due to the refresh rate of the device or some device-specific refresh settings, or limitations in the digitizer or the drivers for it) is making the issue more noticeable. Still, the fact that it is "trigger-able" on other devices means maybe there is some room for improvement? In the sense that there is still some time threshold beneath which successive taps are ignored/interpreted as swipes. I can provide logs is anyone is interested in having a crack figuring out if there is a way to allow quicker typing and not recognizing quick taps as swipe or long-press+special character tap or whatever is happening.

**Comment by Frenzie** [comment]
I'm not sure if it was already mentioned, but there are also various time thresholds you can play with in the menu.

**Comment by hugleo** [comment]
Seems fine here.

![video_2024-04-14_09-36-42-ezgif com-video-to-gif-converter](https://github.com/koreader/koreader/assets/4573505/b932442b-ba59-4d5b-a400-3816f27ea933)

**Comment by tarkeez** [comment]
> I'm not sure if it was already mentioned, but there are also various time thresholds you can play with in the menu.

Thanks. "Tap interval" and "Tap interval on keyboard" settings seem most relevant, and they default to 0 ms. Double tap lower limit is 100 ms, doesn't seem to make a difference.

I tried typing on the regular system kb on the boox page (I only type longer flowing notes in koreader as book annotations), and it does miss some taps, so there is definitely something making this device (and probably other e-ink low refresh devices) worse than a regular phone/android device. Still, the problem is made worse/is more consistent in koreader.

The Page is small enough to dual thumb "phone type" on which allows for quick typing, but alas.

**Comment by Frenzie** [comment]
Double tap is only relevant if you enable double tap. Which I wouldn't because it makes everything slower. Some people don't seem to care but I have this feeling… ;-)

**Comment by Biep** [comment]
I have the same issue on my PocketBook Era, KOReader version 2024.04-134 (and several earlier versions).
And I am very sure I am just tapping.  Most of the time the two taps are too far apart, resulting in two letters missing from the output, but if the second tap is close enough it is interpreted as tapped on the pop-up, and a single special character results.

Unrelated question: are there swipes to get characters from the top row of the pop-up?  The ones two rows above the base character?

**Comment by Frenzie** [comment]
No, those are only accessible with the popup.

**Comment by Biep** [comment]
OK, thanks.
Maybe some day the gesture manager will allow one to define gestures for on the keyboard?
(I'd love to have, say, the up and down arrows swipeable from the left-right ones, or some alternate languages swipeable from the globe key.)

**Comment by Biep** [comment]
My impression is that the spurious pop-up never appears when the first few characters are typed - only when output is four or more characters behind on the input.  And sometimes it appears when typing the last character - then the pup-up is there waiting for a reaction when typing is done.
Could it be some kind of buffer overrrun?

**Comment by Frenzie** [comment]
On which device are you?

**Comment by Biep** [comment]
PocketBook Era.  (See my comment of three days ago.)

**Comment by Frenzie** [comment]
I can only assume there's something too slow/blocking about screen updates on that device then. I'm not sure if anything can be (reasonably) done about that.

**Comment by Biep** [comment]
The screen itself is fast - I can even do smooth scrolling when reading a web page, the jarring is *just* visible, but not breaking the scroll illusion.  Paging forward or backward in KOReader is fast too - a delay of maybe 1/4 second, and then the new page shows at once.
Typing with the keyboard is slow - slower than it was with my Kobo Aura One.  Maybe just under half a second between letters.
(None of this timed strictly.)

**Comment by NiLuJe** [comment]
See https://github.com/koreader/koreader/discussions/11962#discussioncomment-9732159 for my thoughts on the subject.

TL;DR: We can't do precise hold timers on (unrooted) PB.

Note that, even on rooted device, the codepath doesn't trust the kernel timestamp. That's something that could be implemented properly, though, but hardly anyone ever tested that codepath.
(Because it requires a rooted device, to begin with, and a device block with a properly set-up touch translation matrix. That's only ever been done for the *one* device this was developed on, the InkPad 3 Pro).

**Comment by NiLuJe** [comment]
For the affected device, I would try making the hold timer *much* larger as a quick experiment.

Also, verbose debug logs should make the misbehavior slightly more obvious (i.e., you should see holds instead of taps) ;).

**Comment by Biep** [comment]
The problem has gone almost away, after an update of both the PocketBook Era software and KOReader (to 2024.07-29-g761cf1822_2024-07-26).  Only when I type **really** fast, and a fairly long string of characters, it sometimes still brings up the special char pop-up.
Went from barely usable keyboard to virtually perfect overnight.
